example/w-hot-reloading/hot_reloading.ml

```diff
@@ -0,0 +1,89 @@
+let hotreload_script
+    ?(retry_interval_ms = 500)
+    ?(max_retry_ms = 5000)
+    ?(route = "/_hotreload")
+    ()
+  =
+  Printf.sprintf
+    {js|
+var socketUrl = "ws://" + location.host + "%s"
+var s = new WebSocket(socketUrl);
+
+s.onopen = function(even) {
+  console.log("WebSocket connection open.");
+};
+
+s.onclose = function(even) {
+  console.log("WebSocket connection closed.");
+  const innerMs = %i;
+  const maxMs = %i;
+  const maxAttempts = Math.round(maxMs / innerMs);
+  let attempts = 0;
+  function reload() {
+    attempts++;
+    if(attempts > maxAttempts) {
+      console.error("Could not reconnect to dev server.");
+      return;
+    }
+
+    s2 = new WebSocket(socketUrl);
+
+    s2.onerror = function(event) {
+      setTimeout(reload, innerMs);
+    };
+
+    s2.onopen = function(event) {
+      location.reload();
+    };
+  };
+  reload();
+};
+
+s.onerror = function(event) {
+  console.error("WebSocket error observed:", event);
+};
+|js}
+    route
+    retry_interval_ms
+    max_retry_ms
+
+let inject_hotreload_script
+    ?(reload_script = hotreload_script ())
+    ()
+    (next_handler : Dream.request -> Dream.response Lwt.t)
+    (request : Dream.request)
+    : Dream.response Lwt.t
+  =
+  let%lwt response = next_handler request in
+  match Dream.header "Content-Type" response with
+  | Some "text/html" | Some "text/html; charset=utf-8" ->
+    let%lwt body = Dream.body response in
+    let soup = Soup.parse body in
+    let open Soup.Infix in
+    (match soup $? "head" with
```

Comment:
Would be nice to drop the dependency to `lambdasoup`. Would an ad-hoc search for `</head>` work? 🤔

Reply:
An ad-hoc search for `<head>` seems better, as there is less chance to end up in a `<script>` tag or similar, and match a quoted `</head>`. However, not all HTML actually has a literal `<head>` element when transmitted. That includes most of Dream's examples. For instance, in [**`1-hello`**](https://github.com/aantron/dream/tree/master/example/1-hello#readme):

```
let () =
  Dream.run (fun _ ->
    Dream.html "Good morning, world!")
```

The `<head>` tag is automatically inserted by HTML5-conforming parsers when loading such a document, though.

Given that developers are probably likely to start from HTML that doesn't have an explicit `<head>`, it seems like scanning for it will tend to bite them.

Reply:
Right, what we could do also is locate the last `</body>` tag and insert the script before it? I think it would be resilient to scripts containing this tag. And if we don't find it, we can simply add the script at the end of the HTML document?

Reply:
this classic problem. here's merely a suggestion:

- add a document rendering middleware, where
  - m/w accepts a mutable map of `request:document-segments`, e.g. `{ req1ref: Some { head: []  }; req2ref: None }`
- some hot-reloading-aware middleware pushes script datas onto `doc_segments[reqXref].head`
- on middleware exit, `Dream.html <html><head>{% ... map doc_segments to_script_tag %} ...`

essentially emulate closures via a mut ref, then don't expect the leafiest middleware to own the final render. instead, the generally leafier middlewares mutate state ( 😢 ), or accumulate state, and on-pop, the doc middleware actually calls `Dream.html`

next.js uses a `<Head />` component that allows participation in document generation, specifically to avoid cases like this, where a parse and re-generate is needed. it's even more quirky in that the component needs know references to the request of which it's modifying, and i think they plan on ejecting from it. regardless, their flow looks like this:

```
HTTP GET /  ->                    _app m/w   ->   react-render
(stream)        <-  (html doc string)         <- (body string)
```

Reply:
It's possible to to associate HTML document trees with responses using [request-/response-local variables](https://aantron.github.io/dream/#variables), with something like:

```ocaml
let html_tree = Dream.new_local ()

(* In your handler, or a helper to be called from handlers. *)
response
|> Dream.with_local html_tree my_dom_alike

(* Retrieving it in a middleware. *)
match Dream.local html_tree response with
| Some dom_alike -> Dream.html (render dom_alike)
| None -> (* ... *)
```

I'm not sure this is worth it for a hot-reloading middleware &mdash; I assume it would be used in development only, and parsing the HTML again is, AFAIK, not an issue. However, this could be useful in a higher-level framework that needs to defer HTML rendering in production.